File: VB/ClustererSample/Form1.vb
Imports DevExpress.XtraMap
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Windows.Forms
Imports System.Xml.Linq

Namespace ClustererSample

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

#Region "#Clusterer"
        Private ReadOnly Property VectorLayer As VectorItemsLayer
            Get
                Return CType(map.Layers("VectorLayer"), VectorItemsLayer)
            End Get
        End Property

        Private ReadOnly Property DataAdapter As ListSourceDataAdapter
            Get
                Return CType(VectorLayer.Data, ListSourceDataAdapter)
            End Get
        End Property

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            DataAdapter.DataSource = LoadData()
            Dim clusterer As DistanceBasedClusterer = New DistanceBasedClusterer With {.ItemMaxSize = 60, .ItemMinSize = 14, .GroupProvider = New AttributeGroupProvider With {.AttributeName = "LocationName"}}
            clusterer.SetClusterItemFactory(New CustomClusterItemFactory())
            DataAdapter.Clusterer = clusterer
            DataAdapter.PropertyMappings.Add(New MapDotSizeMapping With {.DefaultValue = 14})
            Dim interactiveMode As MouseHoverInteractiveClusterMode = New MouseHoverInteractiveClusterMode()
            interactiveMode.ExpandedClusterLayout = New ExpandedClusterAdaptiveLayout()
            map.InteractiveClusterMode = interactiveMode
        End Sub

#End Region  ' #Clusterer
        Private Function LoadData() As List(Of Tree)
            Dim trees As List(Of Tree) = New List(Of Tree)()
            Dim doc As XDocument = XDocument.Load("Data\TreesCl.xml")
            For Each xTree As XElement In doc.Element("RowSet").Elements("Row")
                trees.Add(New Tree With {.Latitude = Convert.ToDouble(xTree.Element("lat").Value, CultureInfo.InvariantCulture), .Longitude = Convert.ToDouble(xTree.Element("lon").Value, CultureInfo.InvariantCulture), .LocationName = xTree.Element("location").Value.ToString(CultureInfo.InvariantCulture)})
            Next

            Return trees
        End Function
    End Class

#Region "#Factory"
    Friend Class CustomClusterItemFactory
        Inherits IClusterItemFactory

        Public Function CreateClusterItem(ByVal objects As IList(Of MapItem)) As MapItem
            Return New MapDot()
        End Function

        Public Sub CustomizeCluster(ByVal cluster As MapItem)
            CType(cluster, MapDot).TitleOptions.Pattern = cluster.ClusteredItems.Count.ToString()
        End Sub
    End Class

#End Region  ' #Factory
    Friend Class Tree

        Public Property Latitude As Double

        Public Property Longitude As Double

        Public Property LocationName As String
    End Class
End Namespace

File: CS/ClustererSample/Form1.cs
using DevExpress.XtraMap;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Windows.Forms;
using System.Xml.Linq;

namespace ClustererSample {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        #region #Clusterer
        VectorItemsLayer VectorLayer { get { return (VectorItemsLayer)map.Layers["VectorLayer"]; } }
        ListSourceDataAdapter DataAdapter { get { return (ListSourceDataAdapter)VectorLayer.Data; } }

        private void Form1_Load(object sender, EventArgs e) {
            DataAdapter.DataSource = LoadData();
            DistanceBasedClusterer clusterer = new DistanceBasedClusterer {
                ItemMaxSize = 60,
                ItemMinSize = 14,
                GroupProvider = new AttributeGroupProvider {
                    AttributeName = "LocationName"
                }
            };

            clusterer.SetClusterItemFactory(new CustomClusterItemFactory());
            DataAdapter.Clusterer = clusterer;

            DataAdapter.PropertyMappings.Add(new MapDotSizeMapping { DefaultValue = 14 });

            MouseHoverInteractiveClusterMode interactiveMode = new MouseHoverInteractiveClusterMode();
            interactiveMode.ExpandedClusterLayout = new ExpandedClusterAdaptiveLayout();
            map.InteractiveClusterMode = interactiveMode;
        }
        #endregion #Clusterer

        List<Tree> LoadData() {
            List<Tree> trees = new List<Tree>();
            XDocument doc = XDocument.Load("Data\\TreesCl.xml");
            foreach (XElement xTree in doc.Element("RowSet").Elements("Row")) {
                trees.Add(new Tree {
                    Latitude = Convert.ToDouble(xTree.Element("lat").Value, CultureInfo.InvariantCulture),
                    Longitude = Convert.ToDouble(xTree.Element("lon").Value, CultureInfo.InvariantCulture),
                    LocationName = xTree.Element("location").Value.ToString(CultureInfo.InvariantCulture)
                });
            }
            return trees;
        }
    }

    #region #Factory
    class CustomClusterItemFactory : IClusterItemFactory {
        public MapItem CreateClusterItem(IList<MapItem> objects) {
            return new MapDot();
        }

        public void CustomizeCluster(MapItem cluster) {
            ((MapDot)cluster).TitleOptions.Pattern = cluster.ClusteredItems.Count.ToString();
        }
    }
    #endregion #Factory

    class Tree {
        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public string LocationName { get; set; }
    }
}
